docs/tutorial/torch_fsdp_nanogpt.md

```diff
@@ -0,0 +1,144 @@
+# Switching from DDP to FSDP with NanoGPT
+
+Welcome to this guide on how to transition from using DDP (DistributedDataParallel) to FSDP (Fully Sharded Data Parallelism) for training the NanoGPT model. This guide assumes familiarity with the previous DDP guide. If you're new to DDP, we recommend checking out the DDP guide first.
+
+
+
+## What is FSDP?
+
+FSDP is an alternative approach to DDP, designed to improve the efficiency of distributed training. It achieves this by effectively partitioning data and model parameters, reducing communication overhead, and enabling more efficient training on large-scale models.
+
+
+
+## Configuring FSDP for NanoGPT
+
+To replace DDP with FSDP in your existing NanoGPT training configuration, simply make the following changes. Use the `kubectl` command to apply the modified training configuration:
+
+```bash
+$ kubectl -n dlrover apply -f dlrover/examples/torch_nanogpt_job_with_fsdp.yaml
+```
+
+Upon successful application of the job configuration, you can monitor the status of the training nodes using the command below:
+
+```bash
+$ kubectl -n dlrover get pods
+```
+
+
+
+## Comparing DDP and FSDP Results
+
+Let's compare the results obtained using DDP and FSDP with the same parameter settings. Here are the results for the two approaches:
+
+**DDP:**
+
+```bash
+# parameter settings in dlrover/examples/torch_nanogpt_job.yaml
+--n_layer 6 \
+--n_head 6 \
+--n_embd 384
+```
+
+**FSDP (Same Parameter Setting as DDP):**
+
+```bash
+# parameter settings in dlrover/examples/torch_nanogpt_job_with_fsdp.yaml
+--n_layer 6 \
+--n_head 6 \
+--n_embd 384
+```
+
+
+
+### More detailed description of the pods:
+
+Worker-0 Logs
+
+```bash
+
+$ kubectl logs -n dlrover torch-nanogpt-edljob-worker-0
+
+```
+
+results on DDP:
+
+```
+iter 0: loss 4.2516, time 1259.10ms, mfu -100.00%, lr 6.00e-04, total time 1.26s
+iter 1: loss 3.5361, time 26.30ms, mfu -100.00%, lr 6.00e-04, total time 1.29s
```

Comment:
It is better to show the cuda memory of DDP and FSDP. Then, the log can show FSDP can reduce the memory of each node.
```
cuda_mem = torch.cuda.max_memory_allocated()
            print(
                f"iter {iter_num}: loss {lossf:.4f}, time {dt*1000:.2f}ms, "
                f"mfu {running_mfu*100:.2f}%,  cuda memory {cuda_mem / 1e9:.3f}G, "
                f"lr {lr:.2e}, total time {total_time:.2f}s"
            )
```